iam/iam.go

```diff
@@ -126,7 +126,7 @@ func (iam *Client) EndpointFor(service, region string, optFns ...func(*endpoints
 }
 
 // AssumeRole returns an IAM role Credentials using AWS STS.
-func (iam *Client) AssumeRole(roleARN, remoteIP string, sessionTTL time.Duration) (*Credentials, error) {
+func (iam *Client) AssumeRole(roleARN, externalID string, remoteIP string, sessionTTL time.Duration) (*Credentials, error) {
 	hitCache := true
 	item, err := cache.Fetch(roleARN, sessionTTL, func() (interface{}, error) {
```

Comment:
It looks like this is the crux of the caching logic. For this `ccache` get-set mechanism, updating this key to `roleARN:externalID` would help with cache invalidation when an external id is added to an iam role. However it looks like there's no mechanism for invalidating the cache if a role's credentials are invalidated before it's original TTL (invalidation seems like a broader issue).

One other point I'll add, if an iam role requires an external id, there shouldn't be a scenario where two pods are successfully using that role but only one of them is passing in the external id. If Pod A isn't using the external id, sts.AssumeRole requests will fail and nil'ify the cache https://github.com/jtblin/kube2iam/pull/177/files#diff-d40ff635e38cff8f5959411f50e5a8a6R163
meanwhile Pod B is using the external id and is populating the cache with healthy credentials that Pod A could then grab.

TLDRl; I need to update the key to include external id.

Reply:
yeah, I am not sure how you intend to use the external id field, however it would still be possible if two pods with the same role/external id to be co-located on the same host, which would result in a "cache hit" as well.

If you are trying to get "Container Identity" built into the STS role assumption so you know *exactly* which pod is making the request (which could be very valuable in some compliance settings) this still won't address the issue.

If you could outline the "problem statement" of the PR that you are hoping to address that may be helpful ensuring we are aligned on the implications of the changes and how we might architect kube2iam to accommodate those needs.

Reply:
>it would still be possible if two pods with the same role/external id to be co-located on the same host, which would result in a "cache hit" as well.

I think that type of cache hit is acceptable for my use case (for now).

>If you are trying to get "Container Identity" built into the STS role assumption so you know exactly which pod is making the request (which could be very valuable in some compliance settings) this still won't address the issue.

Understood - that's not part of my use case (yet) - but I can see the compliance oriented value. I imagine beyond adding pod name to the cache key could also offer another solution where kube2iam prevents a role from being associated with a pod more than once (or scope it to more than once per namespace and add pod namespace to the cache key). In other words, adding a configuration flag to kube2iam that enforces a 1:1 relationship between either pod:iam-role or namespaced-pod:iam-role.

>If you could outline the "problem statement" of the PR that you are hoping to address that may be helpful ensuring we are aligned on the implications of the changes and how we might architect kube2iam to accommodate those needs.

Will do.

Reply:
Marking this key uniqueness conversation as resolved for now.